File: python/octoarms_sdk/long_text_test.py
from __future__ import annotations

import unittest

from octoarms_sdk.long_text import (
    quote_exists_in_text,
    raise_for_evidence_graph_issues,
    split_text,
    validate_evidence_graph,
)


class LongTextTests(unittest.TestCase):
    def test_split_text_preserves_source_ranges_and_overlap(self) -> None:
        text = "\n\n".join(f"paragraph {index} " + ("x" * 200) for index in range(10))

        chunks = split_text(text, target_chars=700, overlap_chars=100, min_chunk_chars=300)

        self.assertGreater(len(chunks), 1)
        self.assertEqual(chunks[0].chunk_id, "chunk_0000")
        self.assertEqual(chunks[0].index, 0)
        self.assertEqual(chunks[0].char_start, 0)
        self.assertLess(chunks[0].char_end, len(text))
        self.assertLess(chunks[1].char_start, chunks[0].char_end)
        self.assertEqual(chunks[0].text, text[chunks[0].char_start : chunks[0].char_end].strip())

    def test_split_text_rejects_invalid_overlap(self) -> None:
        with self.assertRaisesRegex(ValueError, "overlap_chars"):
            split_text("hello", target_chars=10, overlap_chars=10)

    def test_quote_exists_in_text_normalizes_whitespace_and_case(self) -> None:
        source = "AI factories are\nbecoming a NEW industrial foundation."

        self.assertTrue(quote_exists_in_text("ai factories are becoming a new industrial foundation", source))

    def test_validate_evidence_graph_accepts_consistent_refs(self) -> None:
        issues = validate_evidence_graph(
            insights=[{"insight_id": "ins_1", "evidence_ids": ["ev_1"]}],
            evidence=[{"evidence_id": "ev_1", "quote": "AI factories", "supports": ["ins_1"]}],
            source_text="Jensen said AI factories are industrial infrastructure.",
            validate_quotes_in_source=True,
        )

        self.assertEqual(issues, [])

    def test_validate_evidence_graph_reports_dangling_refs_and_missing_reverse_support(self) -> None:
        issues = validate_evidence_graph(
            insights=[{"insight_id": "ins_1", "evidence_ids": ["ev_1", "ev_missing"]}],
            evidence=[{"evidence_id": "ev_1", "quote": "AI factories", "supports": ["other"]}],
        )

        self.assertEqual([issue.code for issue in issues], ["missing_reverse_support", "dangling_evidence_ref"])
        with self.assertRaisesRegex(ValueError, "missing_reverse_support"):
            raise_for_evidence_graph_issues(issues)

    def test_validate_evidence_graph_reports_quote_not_in_source(self) -> None:
        issues = validate_evidence_graph(
            insights=[{"insight_id": "ins_1", "evidence_ids": ["ev_1"]}],
            evidence=[{"evidence_id": "ev_1", "quote": "not present", "supports": ["ins_1"]}],
            source_text="AI factories are present.",
            validate_quotes_in_source=True,
        )

        self.assertEqual(issues[0].code, "quote_not_in_source")


if __name__ == "__main__":
    unittest.main()

File: typescript/src/long_text.test.ts
import assert from "node:assert/strict"
import { describe, it } from "node:test"

import { quoteExistsInText, raiseForEvidenceGraphIssues, splitText, validateEvidenceGraph } from "./long_text"

describe("long text helpers", () => {
  it("splits text with source ranges and overlap", () => {
    const text = Array.from({ length: 10 }, (_, index) => `paragraph ${index} ${"x".repeat(200)}`).join("\n\n")

    const chunks = splitText(text, { targetChars: 700, overlapChars: 100, minChunkChars: 300 })

    assert.ok(chunks.length > 1)
    assert.equal(chunks[0].chunkId, "chunk_0000")
    assert.equal(chunks[0].index, 0)
    assert.equal(chunks[0].charStart, 0)
    assert.ok(chunks[0].charEnd < text.length)
    assert.ok(chunks[1].charStart < chunks[0].charEnd)
    assert.equal(chunks[0].text, text.slice(chunks[0].charStart, chunks[0].charEnd).trim())
  })

  it("rejects invalid overlap", () => {
    assert.throws(() => splitText("hello", { targetChars: 10, overlapChars: 10 }), /overlapChars/)
  })

  it("normalizes whitespace and case when checking quotes", () => {
    const source = "AI factories are\nbecoming a NEW industrial foundation."

    assert.equal(quoteExistsInText("ai factories are becoming a new industrial foundation", source), true)
  })

  it("accepts consistent evidence refs", () => {
    const issues = validateEvidenceGraph(
      [{ insight_id: "ins_1", evidence_ids: ["ev_1"] }],
      [{ evidence_id: "ev_1", quote: "AI factories", supports: ["ins_1"] }],
      { sourceText: "Jensen said AI factories are industrial infrastructure.", validateQuotesInSource: true },
    )

    assert.deepEqual(issues, [])
  })

  it("reports dangling refs and missing reverse support", () => {
    const issues = validateEvidenceGraph(
      [{ insight_id: "ins_1", evidence_ids: ["ev_1", "ev_missing"] }],
      [{ evidence_id: "ev_1", quote: "AI factories", supports: ["other"] }],
    )

    assert.deepEqual(
      issues.map((issue) => issue.code),
      ["missing_reverse_support", "dangling_evidence_ref"],
    )
    assert.throws(() => raiseForEvidenceGraphIssues(issues), /missing_reverse_support/)
  })

  it("reports quote not found in source", () => {
    const issues = validateEvidenceGraph(
      [{ insight_id: "ins_1", evidence_ids: ["ev_1"] }],
      [{ evidence_id: "ev_1", quote: "not present", supports: ["ins_1"] }],
      { sourceText: "AI factories are present.", validateQuotesInSource: true },
    )

    assert.equal(issues[0].code, "quote_not_in_source")
  })
})
